b16set/b16set.go

```diff
@@ -145,100 +133,118 @@ func (s *Set) Each(f func(item [16]byte) bool) {
 
 // Copy returns a new Set with a copy of s.
 func (s *Set) Copy() *Set {
-	u := New()
+	u := &Set{
+		m: make(map[[16]byte]struct{}, s.Size()),
+	}
 	for item := range s.m {
-		u.Add(item)
+		u.m[item] = keyExists
 	}
 	return u
 }
 
 // String returns a string representation of s
 func (s *Set) String() string {
-	t := make([]string, 0, len(s.List()))
-	for _, item := range s.List() {
-		t = append(t, fmt.Sprintf("%v", item))
+	v := make([]string, 0, s.Size())
+	for item := range s.m {
+		v = append(v, fmt.Sprintf("%v", item))
 	}
-
-	return fmt.Sprintf("[%s]", strings.Join(t, ", "))
+	return fmt.Sprintf("[%s]", strings.Join(v, ", "))
 }
 
 // List returns a slice of all items. There is also StringSlice() and
 // IntSlice() methods for returning slices of type string or int.
 func (s *Set) List() [][16]byte {
-	list := make([][16]byte, 0, len(s.m))
-
+	v := make([][16]byte, 0, s.Size())
 	for item := range s.m {
-		list = append(list, item)
+		v = append(v, item)
 	}
-
-	return list
+	return v
 }
 
 // Merge is like Union, however it modifies the current Set it's applied on
 // with the given t Set.
 func (s *Set) Merge(t *Set) {
-	t.Each(func(item [16]byte) bool {
+	for item := range t.m {
 		s.m[item] = keyExists
-		return true
-	})
+	}
 }
 
 // Separate removes the Set items containing in t from Set s. Please aware that
 // it's not the opposite of Merge.
 func (s *Set) Separate(t *Set) {
-	s.Remove(t.List()...)
+	for item := range t.m {
+		delete(s.m, item)
+	}
 }
 
 // Union is the merger of multiple sets. It returns a new set with all the
 // elements present in all the sets that are passed.
-func Union(set1, set2 *Set, sets ...*Set) *Set {
-	u := set1.Copy()
-	set2.Each(func(item [16]byte) bool {
-		u.Add(item)
-		return true
-	})
-	for _, set := range sets {
-		set.Each(func(item [16]byte) bool {
-			u.Add(item)
-			return true
-		})
+func Union(sets ...*Set) *Set {
+	maxPos := -1
```

Comment:
Shouldn't maxPos be set to 0 here?

Reply:
I guess it's covered by the next check.

Reply:
0 is not good either `if len(sets) == 0`

Reply:
But the check after catches this so I guess it's fine.